### DISCUSSION TITLE
Using VS Code/rust-analyzer with cargo 3ds

### DISCUSSION THREAD

**Discussion by DeltaF1** [rust3ds/ctru-rs#127] [Q&A] [answered]
I've noticed that when rust-analyzer runs it seems to re-build std every time, and then when I go to build/run on the command line with `cargo 3ds`, it seems to have invalidated the cached std and must re-build yet again. This happens for projects using ctru-rs even when I haven't modified any of ctru-rs itself.

My settings.json in vscode contains these flags to pass to rust-analyzer

```json
    "rust-analyzer.cargo.target": "armv6k-nintendo-3ds",
    "rust-analyzer.cargo.extraArgs": [
        "-Zbuild-std"
    ]
 ```
 
Is anyone else using vscode and can share their settings?

**Comment by Meziu** [comment, marked as answer]
This behaviour also happens when normally executing this build command (rather than `cargo 3ds build`):
`cargo build --target armv6k-nintendo-3ds -Zbuild-std --lib`

You just need to export the env configuration you can see around [here](https://github.com/rust3ds/cargo-3ds/blob/78a652fdfb01e2614a792d1a56b10c980ee1dae9/src/lib.rs#L59).

In simpler terms, add `RUSTFLAGS="-L/opt/devkitpro/libctru/lib -lctru"` to your env.
```json
{
  "rust-analyzer.cargo.extraEnv": { "RUSTFLAGS": "-L/opt/devkitpro/libctru/lib -lctru" }
}
```

As any good programmer would, I don't use this, because I much prefer having `rust-analyzer` spit wrong errors than having to use a configuration file for VSCode :P

**Reply by DeltaF1** [reply]
That worked perfectly, thank you!

**Comment by adryzz** [comment]
I actually have been using rust-analyzer and codelldb for quite a while with this project.

All i had to do was change the "check" command in the settings over to "cargo 3ds check" (remember to set the output to json too). You can even change the command per-project as to not have to re-change it every time you interact with a non-3ds rust project.

`.vscode/settings.json`
```json
{
    "rust-analyzer.check.overrideCommand": [
    
    "cargo",
    "3ds",
    "check",
    "--message-format=json"
],
}
```

As for running, i have a launch.json like [this](https://github.com/adryzz/inochi2d-3ds/blob/master/.vscode/launch.json) that uses codelldb and the Luma3DS GDB server.
